server/route.go

```diff
@@ -1748,6 +1775,32 @@ func (c *client) processRouteConnect(srv *Server, arg []byte, lang string) error
 		perms = srv.getOpts().Cluster.Permissions
 	}
 
+	clusterName := srv.ClusterName()
+
+	// If we have a cluster name set, make sure it matches ours.
+	if proto.Cluster != clusterName {
+		shouldReject := true
+		// If we have a dynamic name we will do additional checks.
+		if srv.getOpts().Cluster.Name == "" {
+			if !proto.Dynamic || strings.Compare(clusterName, proto.Cluster) < 0 {
+				// We will take on their name since theirs is configured or higher then ours.
+				srv.setClusterName(proto.Cluster)
+				if !proto.Dynamic {
+					srv.getOpts().Cluster.Name = proto.Cluster
```

Comment:
If we were to store the "dynamic" nature of the cluster name in the server, we could do that as part of srv.setClusterName() where we would pass the proto.Dynamic boolean to decide if we change our view of dynamic. Changing the option here seem a bit hacky.

Reply:
I went back and forth on this as well with special prefixes etc. If the option is set it is not dynamic felt like the best option given all of the moving pieces.